File: python/Binary Tree Preorder Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        self.ans = []
        self.doTraversal(root)
        return self.ans

    def doTraversal(self, t):
        if t == None:
            return None
        else:
            self.ans.append(t.val)
        if t.left != None:
            self.doTraversal(t.left)
        if t.right != None:
            self.doTraversal(t.right)
        return None

File: src/Binary Tree Preorder Traversal.cpp
#include <vector>
#include <iostream>
using namespace std;

//    Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
 
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        ans.clear();
        doTraversal(root);
        return ans;
    }

    void doTraversal(TreeNode *t) {
        if(t)
            ans.push_back(t->val);
        else
            return;
        if(t->left)
            doTraversal(t->left);
        if(t->right)
            doTraversal(t->right);
    }
private:
    vector<int> ans;
};

int main()
{
    Solution so;
    vector<int> ans;
    ans = so.preorderTraversal(NULL);
    for(int i=0; i<ans.size(); i++)
        cout<<ans[i]<<endl;
}